federationapi/queue/destinationqueue.go

```diff
@@ -388,26 +388,49 @@ func (oq *destinationQueue) backgroundSend() {
 				return
 			}
 		} else {
-			oq.handleTransactionSuccess(pduCount, eduCount)
+			oq.handleTransactionSuccess(pduCount, eduCount, asyncSuccess)
 		}
 	}
 }
 
 // nextTransaction creates a new transaction from the pending event
 // queue and sends it.
-// Returns an error if the transaction wasn't sent.
+// Returns an error if the transaction wasn't sent. And whether the success
+// was to an async relay server or not.
 func (oq *destinationQueue) nextTransaction(
 	pdus []*queuedPDU,
 	edus []*queuedEDU,
-) error {
+) (err error, asyncSuccess bool) {
 	// Create the transaction.
 	t, pduReceipts, eduReceipts := oq.createTransaction(pdus, edus)
 	logrus.WithField("server_name", oq.destination).Debugf("Sending transaction %q containing %d PDUs, %d EDUs", t.TransactionID, len(t.PDUs), len(t.EDUs))
 
 	// Try to send the transaction to the destination server.
 	ctx, cancel := context.WithTimeout(oq.process.Context(), time.Minute*5)
 	defer cancel()
-	_, err := oq.client.SendTransaction(ctx, t)
+
+	relayServers := oq.statistics.KnownRelayServers()
+	if oq.statistics.AssumedOffline() && len(relayServers) > 0 {
+		logrus.Infof("Sending to relay servers: %v", relayServers)
+		// TODO : how to pass through actual userID here?!?!?!?!
+		userID, userErr := gomatrixserverlib.NewUserID("@user:"+string(oq.destination), false)
```

Comment:
It could be reduced to `destinationServer` instead of user.
I did it this way assuming that a relay server may not want to relay transactions for every user associated with a given server. So the relay server could perform some kind of user filtering at this point. 
In general I tried to keep relay servers abstract from the concept of p2p. They functional exactly the same for non-p2p matrix. Whether there is a use case for that I don't know. But I didn't want to bake in p2p assumptions anywhere I didn't need to.